File: Quick_Sort.cpp
#include<iostream>
using namespace std;

int partition(int arr[],int low,int high){
    int temp;
    int pivot=arr[low];
    int i=low+1;
    int j=high;

    do{
    while(arr[i]<=pivot){
        i++;
    }
    while(arr[j]>pivot){
        j--;
    }
    if(i<j){
    temp=arr[i];
    arr[i]=arr[j];
    arr[j]=temp;
    }
    }while(i<j);
    temp=arr[low];
    arr[low]=arr[j];
    arr[j]=temp;
    return j;
}

void Quick_Sort(int arr[],int low,int high){
    int partitionIndex;
    if(low<high){
    partitionIndex = partition(arr,low,high);
    Quick_Sort(arr, low, partitionIndex-1);
    Quick_Sort(arr,partitionIndex+1,high);
    }


}

void printArray(int arr[],int n){
    for(int i=0;i<n;i++){
        cout<<arr[i]<<" ";
    }
    cout<<endl;
}


int main(){
    int n;
    cout<<"Enter Size Of Array:";
    cin>>n;
    int arr[n];
    cout<<"Enter Array Elements:";
    for(int i=0;i<n;i++){
        cin>>arr[i];
    }
    printArray(arr,n);
    Quick_Sort(arr,0,n-1);
    printArray(arr,n);
    return 0;
}

File: Recursion/QuickSort.java


public class QuickSort {

    public static int swap(int[] arr, int low, int high) {
        int pivot = arr[low];
        int i = low, j = high;
        while (i < j) {
            while (arr[i] <= pivot && i <= high) {
                i++;
            }
            while (arr[j] > pivot && j >= low) {
                j--;
            }
            if (i < j) {
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }

            int temp = arr[low];
            arr[low] = arr[j];
            arr[j] = temp;
        }
        return j;
    }

    public static void qs(int[] arr, int low, int high) {
        if (low < high) {
            int partition = swap(arr, low, high);
            qs(arr, low, partition - 1);
            qs(arr, partition + 1, high);
        }
    }

    public static void main(String[] args) {
        int[] arr = {3, 1, 5, 4, 2};
        int low = 0;
        int high = arr.length - 1;
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i]);
        }
        System.err.println("");
        qs(arr, low, high);
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i]);
        }
    }
}
